Create a guide to using the Compiler locally (#4516)
VATSIM-UK/UK-Sector-File: .github/Compiler Guide.md

# Using the Sector File Compiler

## Introduction

The Sector File Compiler is the open-source tool that is used to transform the files that make up the Sector File repository into the .sct, .ese, and .rwy files that Euroscope uses. The link to the repository is [here](https://github.com/VATSIM-UK/sector-file-compiler), and if you should have any issues with the Compiler not working as intended (a verifiable bug), or suggestions to improve it, then do feel free to leave an issue on that repository. If you have questions about its use or aren't quite sure how to do something, do put a message in the #sector_file_development channel in the VATSIM UK Discord.

The Compiler, aside from just merely putting the files into the correct places in the Sector File, is also capable of validating input. This makes it incredibly useful, bordering on essential, for developing 'correct' sector files that work as intended and catching very understandable human error when dealing with Euroscope's esoteric syntax requirements. For example, coordinates entered in the wrong format will produce the following message:

> ![image](https://user-images.githubusercontent.com/14115426/208726651-1ff76baa-a197-4120-acb6-1e9f7061d013.png)

Reviewer: I personally found this image a bit too wide to view properly in the markdown viewer thing I was using, I wonder if might be better to have as a code quote type thing? But upto you which you think is best.

Reviewer: Same image appears at bottom too I think, so would apply there too if you wanted to change.

Author: Remade it - let me know if it still seems to wide

Reviewer: That's better, thanks.

This guide will make the following assumptions, as these seem to be the most common circumstances under which contributors work:

- You are using Windows 10/11
- You are using GitHub Desktop

It is very much possible to use the Compiler under completely contrary circumstances, i.e. you are not using Windows 10/11, and not using GitHub desktop (e.g. using Git standalone). If this applies, and would like guidance on how to use the Compiler, do put a message in #sector_file_development, as it is likely someone can help. On the other hand, if you are happy using the compiler, then the author would be grateful if you could extend this guide for those circumstances through a PR - every contribution helps!

## Downloading the Compiler

> To download the compiler, first navigate to the [Compiler releases page](https://github.com/VATSIM-UK/sector-file-compiler/releases). Then, download the latest release - marked with a green label of 'Latest and will be at the top of the list. You will want to click the 'cli-windows-x64' button - this should trigger a download:

Reviewer: ```suggestion
To download the compiler, first navigate to the [Compiler releases page](https://github.com/VATSIM-UK/sector-file-compiler/releases). Then, download the latest release - marked with a green label of 'Latest' and will be at the top of the list. You will want to click the 'cli-windows-x64' button - this should trigger a download:

```

![image](https://user-images.githubusercontent.com/14115426/208726535-de2d20d0-0b00-45fd-b2f8-9d79974dc850.png)

> Once this has downloaded, open your Downloads folder, and rename the 'cli-windows-x64' file to 'cli-windows-x64.exe'. This allows Windows to run it. It then needs to be moved by cutting-and-pasting it into your local repository folder. If you are using GitHub Desktop, this will be in your Documents->GitHub->UK-Sector-File. It needs to go into that final folder, i.e. Documents/GitHub/UK-Sector-File/cli-compiler-x64.exe. Don't worry about Git picking this up - all .EXEs are automatically excluded, as are the filenames.

Reviewer: Mine is somewhere else, I think you can choose where you want it when you set up the fork, I have it in `C:\Users\peter\OneDrive\Documents\Flying\VATSIM\UK SF Dev\GitHub_UK-Sector-File`.
```suggestion
Once this has downloaded, open your Downloads folder, and rename the 'cli-windows-x64' file to 'cli-windows-x64.exe'. This allows Windows to run it. It then needs to be moved by cutting-and-pasting it into your local repository folder. If you are using GitHub Desktop, this will probably be in your Documents->GitHub->UK-Sector-File. It needs to go into that final folder, e.g. Documents/GitHub/UK-Sector-File/cli-compiler-x64.exe. Don't worry about Git picking this up - all .EXEs are automatically excluded, as are the filenames.

```

## Basic usage of the Compiler

The most basic command to run the compiler, generating the Sector File and running the associated validations, is: `./cli-windows-x64.exe --config-file compiler.config.json`

The generated files can then be found in the .bin folder (Documents/GitHub/UK-Sector-File/.bin). These can be loaded in Euroscope through the normal Load Sector File menu. The .bin folder is also excluded from git and so will not be accidentally included in PRs.

The `--config-file` flag sets the config file (the file that the compiler uses to decide which files to include and how), and this should in almost all cases be just `compiler.config.json`, which is the file the repo uses. If you are developing a new config, this can be changed by simply entering the name of the new file you wish to use instead.

If all you wish to do is use the compiler to generate the Sector File locally, then you are done! A few things to bear in mind:

- The command must be run after each set of changes have been made (it does not run automatically)

> - You must reload the sector file each time you generate a new one (as it is loaded completely into memory and does not re-read it)

Reviewer: ```suggestion
- You must reload the sector file each time you generate a new one (as it is loaded completely into memory and Euroscope does not re-read it without being prompted)

```
